api/src/main/java/io/opentelemetry/trace/Span.java

```diff
@@ -217,6 +217,29 @@
    */
   void setStatus(Status status);
 
+  /**
+   * Records information about the {@link Throwable} to the {@link Span} as an {@link Event} at the
+   * current time.
+   *
+   * @param exception the {@link Throwable} to record.
+   * @since 0.7.0
+   */
+  void recordException(Throwable exception);
+
+  /**
+   * Records information about the {@link Throwable} to the {@link Span} as an {@link Event}.
+   *
+   * <p>Use this method to specify an explicit event timestamp. If not called, the implementation
+   * will use the current timestamp value, which should be the default case.
+   *
+   * <p>Important: this is NOT equivalent with System.nanoTime().
+   *
+   * @param exception the {@link Throwable} to record.
+   * @param timestamp the explicit event timestamp in nanos since epoch.
+   * @since 0.7.0
+   */
+  void recordException(Throwable exception, long timestamp);
```

Comment:
Would not add this method unless is necessary. Removing methods is harder than adding. I don't see a scenario when an exception happens and somehow later want to record it.

Reply:
I would argue anything with a timestamp needs to have a way to manually set it, because users may have their own clocks, or some similar reason where timing is kept track in their framework and they want to use it. I can remove this method but then I'd argue it's a pre-GA task to reconsider modeling exceptions as events since it means we don't consider the timestamp to be meaningful. @iNikem any thoughts?

Reply:
Actually user using their own clock is the main reason I am against this :). Reason is we do a good job using nanoTime to ensure ordering of the events, but user does not have access to our internal clock so any time recorded by external user will not be order with the other events.

Reply:
I don't think this is correct in a general sense, many frameworks record timings at a level too deep for us to capture as accurately, at least without bytecode instrumentation. For example a "wire send" event, if it's recorded just before passing a buffer to netty will probably be many nanoseconds more accurate than what our instrumentation could pull off. Note that in case, user is interchangable with instrumentation, the latter also needs to be able to pass on the accurate timings. There is some concern about ordering issues with non-framework generated events, but I'd take the more accurate timings for things like DNS resolution, etc any day.

But I agree this seems far less useful for exceptions. So it makes me wonder whether we really need a timestamp for exceptions in the first place if, similar to a DNS query, we may miss out on a better timing provided by the framework if we choose not to allow that here.

In the meantime, I've removed the method though as it seems like a bigger discussion about why exception is modeled as Event instead of something without a timestamp.

Reply:
I took a look at [API defined by spec](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#add-events). 

```
An Event is defined by the following properties:

(Required) Name of the event.
(Optional) One or more Attributes with the same restrictions as defined for Span Attributes.
(Optional) Timestamp for the event.
```

Several other API methods accept timestamp as well. Thus this method is totally within spec. Although I too have doubts if anyone will use it.

Reply:
Thanks for the link to the spec! I didn't check it, just looked at our Java API. I feel our current API doesn't reflect the optional aspect of the Timestamp (at least the javadoc needs to be more explicit about using current time when not provided and pointing at a way to create an event without timestamps). In that case, I'm good with this API I'll investigate this optional-timestamp event case more generally in a separate PR.

Reply:
Filed #1447 to track that

Reply:
Ah realized my comment was ambiguous - I meant I'm good with not including the method version that includes a timestamp. As part of #1447 I'll probably switch this to add an event without any timestamp.